{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext sql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: DATABASE_URL=postgresql://itversity_retail_user:itversity@localhost:5432/itversity_retail_db\n"
     ]
    }
   ],
   "source": [
    "%env DATABASE_URL=postgresql://itversity_retail_user:itversity@localhost:5432/itversity_retail_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'postgresql://itversity_retail_user:itversity@localhost:5432/itversity_retail_db'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%env DATABASE_URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/base.py\", line 146, in __init__\n",
      "    self._dbapi_connection = engine.raw_connection()\n",
      "                             ~~~~~~~~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/base.py\", line 3298, in raw_connection\n",
      "    return self.pool.connect()\n",
      "           ~~~~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 449, in connect\n",
      "    return _ConnectionFairy._checkout(self)\n",
      "           ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 1263, in _checkout\n",
      "    fairy = _ConnectionRecord.checkout(pool)\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 712, in checkout\n",
      "    rec = pool._do_get()\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/impl.py\", line 179, in _do_get\n",
      "    with util.safe_reraise():\n",
      "         ~~~~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/util/langhelpers.py\", line 146, in __exit__\n",
      "    raise exc_value.with_traceback(exc_tb)\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/impl.py\", line 177, in _do_get\n",
      "    return self._create_connection()\n",
      "           ~~~~~~~~~~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 390, in _create_connection\n",
      "    return _ConnectionRecord(self)\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 674, in __init__\n",
      "    self.__connect()\n",
      "    ~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 900, in __connect\n",
      "    with util.safe_reraise():\n",
      "         ~~~~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/util/langhelpers.py\", line 146, in __exit__\n",
      "    raise exc_value.with_traceback(exc_tb)\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 896, in __connect\n",
      "    self.dbapi_connection = connection = pool._invoke_creator(self)\n",
      "                                         ~~~~~~~~~~~~~~~~~~~~^^^^^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/create.py\", line 646, in connect\n",
      "    return dialect.connect(*cargs, **cparams)\n",
      "           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/default.py\", line 622, in connect\n",
      "    return self.loaded_dbapi.connect(*cargs, **cparams)\n",
      "           ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/psycopg2/__init__.py\", line 122, in connect\n",
      "    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)\n",
      "psycopg2.OperationalError: connection to server at \"localhost\" (::1), port 5432 failed: Connection refused\n",
      "\tIs the server running on that host and accepting TCP/IP connections?\n",
      "connection to server at \"localhost\" (127.0.0.1), port 5432 failed: Connection refused\n",
      "\tIs the server running on that host and accepting TCP/IP connections?\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sql/magic.py\", line 196, in execute\n",
      "    conn = sql.connection.Connection.set(\n",
      "        connect_str,\n",
      "    ...<2 lines>...\n",
      "        creator=args.creator,\n",
      "    )\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sql/connection.py\", line 78, in set\n",
      "    cls.current = Connection(\n",
      "                  ~~~~~~~~~~^\n",
      "        os.getenv(\"DATABASE_URL\"), connect_args, creator\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    )\n",
      "    ^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sql/connection.py\", line 55, in __init__\n",
      "    self.internal_connection = engine.connect()\n",
      "                               ~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/base.py\", line 3274, in connect\n",
      "    return self._connection_cls(self)\n",
      "           ~~~~~~~~~~~~~~~~~~~~^^^^^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/base.py\", line 148, in __init__\n",
      "    Connection._handle_dbapi_exception_noconnection(\n",
      "    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^\n",
      "        err, dialect, engine\n",
      "        ^^^^^^^^^^^^^^^^^^^^\n",
      "    )\n",
      "    ^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/base.py\", line 2439, in _handle_dbapi_exception_noconnection\n",
      "    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/base.py\", line 146, in __init__\n",
      "    self._dbapi_connection = engine.raw_connection()\n",
      "                             ~~~~~~~~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/base.py\", line 3298, in raw_connection\n",
      "    return self.pool.connect()\n",
      "           ~~~~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 449, in connect\n",
      "    return _ConnectionFairy._checkout(self)\n",
      "           ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 1263, in _checkout\n",
      "    fairy = _ConnectionRecord.checkout(pool)\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 712, in checkout\n",
      "    rec = pool._do_get()\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/impl.py\", line 179, in _do_get\n",
      "    with util.safe_reraise():\n",
      "         ~~~~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/util/langhelpers.py\", line 146, in __exit__\n",
      "    raise exc_value.with_traceback(exc_tb)\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/impl.py\", line 177, in _do_get\n",
      "    return self._create_connection()\n",
      "           ~~~~~~~~~~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 390, in _create_connection\n",
      "    return _ConnectionRecord(self)\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 674, in __init__\n",
      "    self.__connect()\n",
      "    ~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 900, in __connect\n",
      "    with util.safe_reraise():\n",
      "         ~~~~~~~~~~~~~~~~~^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/util/langhelpers.py\", line 146, in __exit__\n",
      "    raise exc_value.with_traceback(exc_tb)\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/pool/base.py\", line 896, in __connect\n",
      "    self.dbapi_connection = connection = pool._invoke_creator(self)\n",
      "                                         ~~~~~~~~~~~~~~~~~~~~^^^^^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/create.py\", line 646, in connect\n",
      "    return dialect.connect(*cargs, **cparams)\n",
      "           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/sqlalchemy/engine/default.py\", line 622, in connect\n",
      "    return self.loaded_dbapi.connect(*cargs, **cparams)\n",
      "           ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/woodzsan/Desktop/Machine Learning and Data Analysis/Data-Engineering-Essentials/035 Python Essentials for Data Engineers/06 Files to Database Migration/files-to-database-migration-app/ftd_migration/lib/python3.13/site-packages/psycopg2/__init__.py\", line 122, in connect\n",
      "    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)\n",
      "sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at \"localhost\" (::1), port 5432 failed: Connection refused\n",
      "\tIs the server running on that host and accepting TCP/IP connections?\n",
      "connection to server at \"localhost\" (127.0.0.1), port 5432 failed: Connection refused\n",
      "\tIs the server running on that host and accepting TCP/IP connections?\n",
      "\n",
      "(Background on this error at: https://sqlalche.me/e/20/e3q8)\n",
      "\n",
      "Connection info needed in SQLAlchemy format, example:\n",
      "               postgresql://username:password@hostname/dbname\n",
      "               or an existing connection: dict_keys([])\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "SELECT * FROM orders LIMIT 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ftd_migration",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
